feat(linter)!: support plugins in oxlint config files

DonIsaac opened the pull request:
> Closes #5046
This PR migrates the linter crate and oxlint app to use the new `LinterBuilder` API. This PR has the following effects:

1. `plugins` in oxlint config files are now supported
2. irons out weirdness when using CLI args and config files together. CLI args are now always applied on top of config file settings, overriding them.

# Breaking Changes
Before, config files would override rules set in CLI arguments. For example, running this command:
```sh
oxlint -A correctness -c oxlintrc.json
```
With this config file
```jsonc
// oxlintrc.json
{
  "rules": {
    "no-const-assign": "error"
  }
}
```
Would result in a single rule, `no-const-assign` being turned on at an error level with all other rules disabled (i.e. set to "allow").

Now, **CLI arguments will override config files**. That same command with the same config file will result with **all rules being disabled**.

## Details

For a more in-depth explanation, assume we are running the below command using the `oxlintrc.json` file above:
```sh
oxlint -A all -W correctness -A all -W suspicious -c oxlintrc.json
```

### Before
> Note: GitHub doesn't seem to like deeply nested lists. Apologies for the formatting.

Here was the config resolution process _before_ this PR:
<details><summary>Before Steps</summary>

1. Start with a default set of filters (["correctness", "warn"]) if no filters were passed to the CLI. Since some were, the filter list starts out empty.
2. Apply each filter taken from the CLI from left to right. When a filter allows a rule or category, it clears the configured set of rules. So applying those filters looks like this
  a. start with an empty list `[]`
  b. `("all", "allow")` -> `[]`
  c. `("correctness", "warn")` -> `[ <correctness rules> ]`
  d. `("all", "allow")` -> `[]`
  e. `("suspicious", "warn")` -> `[ <suspicious rules> ]`. This is the final rule set for this step
3. Apply overrides from `oxlintrc.json`. This is where things get a little funky, as mentioned in point 2 of what this PR does. At this point, all rules in the rules list are only from the CLI.
  a. If a rule is only set in the CLI and is not present in the config file, there's no effect
  b. If a rule is in the config file but not the CLI, it gets inserted into the list.
  c. If a rule is already in the list and in the config file
    i. If the rule is only present once (e.g. `"no-loss-of-precision": "error"`), unconditionally override whatever was in the CLI with what was set in the config file
    ii. If the rule is present twice (e.g. `"no-loss-of-precision": "off", "@typescript-eslint/no-loss-of-precision": "error"`),
      a. if all rules in the config file are set to `allow`, then turn the rule off
      b. If one of them is `warn` or `deny`, then update the currently-set rule's config object, but _leave its severity alone_.

  So, for our example, the final rule set would be `[<all suspicious rules: "warn">, no-const-assign: "error"]`

</details>

### After
Rule filters were completely re-worked in a previous PR. Now, lint filters aren't kept on hand-only the rule list is.

<details><summary>After Steps</summary>

1. Start with the default rule set, which is all correctness rules for all enabled plugins (`[<all correctness rules: "warn">]`)
2. Apply configuration from `oxlintrc.json` _first_.
  a. If the rule is warn/deny exists in the list already, update its severity and config object. If it's not in the list, add it.
  b. If the rule is set to allow, remove it from the list.

  The list is now `[<all correctness rules except no-const-assign: "warn">, no-const-assign: "error"]`.

3. Apply each filter taken from the CLI from left to right. This works the same way as before. So, after they're applied, the list is now `[<suspicous rules: "warn">]`

</details>

DonIsaac commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6088?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6088** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6088?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6087** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6087?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6086** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6086?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6085** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6085?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6081** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6081?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6080** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6080?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @DonIsaac and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

DonIsaac commented:
> ## [CodSpeed Performance Report](https://codspeed.io/oxc-project/oxc/branches/don/09-26-feat_linter_support_plugins_in_oxlint_config_files)
> ### Merging #6088 will **degrade performances by 3.72%**
> Comparing `don/09-26-feat_linter_support_plugins_in_oxlint_config_files` ([d08cf2c](https://github.com/oxc-project/oxc/commit/d08cf2cd9deb2af5bdf5413e4c20497aa5061c96)) with `don/09-26-fix_linter_rule_and_generic_filters_do_not_re-configure_existing_rules` ([6edcd5c](https://github.com/oxc-project/oxc/commit/6edcd5ceaab25547e1d5cbbce6ceb44d4f251db5))
> 
> ### Summary
> `❌ 1` regressions `✅ 28` untouched benchmarks
> 
> > ⚠️ _Please fix the performance issues or [acknowledge them on CodSpeed](https://codspeed.io/oxc-project/oxc/branches/don/09-26-feat_linter_support_plugins_in_oxlint_config_files)._
> 
> ### Benchmarks breakdown
> Benchmark	`don/09-26-fix_linter_rule_and_generic_filters_do_not_re-configure_existing_rules`	`don/09-26-feat_linter_support_plugins_in_oxlint_config_files`	Change
> ❌	`linter[checker.ts]`	2.4 s	2.5 s	-3.72%

It looks like some rules are not running that weren't before. This either means this PR fixed a bug that caused them not to run, or ...run() is now taking longer, so those spans are showing up?

TBH I'm not quite sure what to make of this.

camchenry commented on apps/oxlint/src/command/lint.rs:
```diff
@@ -211,64 +211,178 @@ impl FromStr for OutputFormat {
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq)]
+#[allow(clippy::enum_variant_names)]
+pub enum Enable {
```
I'm finding this name a bit confusing, maybe we could differentiate from the `Enable::Enable` member? perhaps: `Enablement`?

DonIsaac replied:
Thanks for the feedback. I'll rename it in the morning.

DonIsaac replied:
I renamed it to `OverrideToggle` and added some doc comments.

camchenry commented on apps/oxlint/src/command/lint.rs:
```diff
@@ -211,64 +211,178 @@ impl FromStr for OutputFormat {
+        matches!(self, Self::Enable)
+    }
+
+    #[inline]
+    pub fn is_not_set(self) -> bool {
+        matches!(self, Self::NotSet)
+    }
```
[nonblocking, feel free to resolve] should we add a `is_set` API as well to check if a given plugin has been configured? might not be needed now though.

camchenry commented on apps/oxlint/src/command/lint.rs:
```diff
@@ -211,64 +211,178 @@ impl FromStr for OutputFormat {
+        matches!(self, Self::NotSet)
+    }
+
+    pub fn inspect<F>(self, f: F)
+    where
+        F: FnOnce(bool),
+    {
```
should this return something? to be consistent w/ Rust I guess it should return `Enable`?

DonIsaac replied:
No. It's based of `Option::inspect`.

camchenry replied:
it looks like `Option::inspect` returns `Self` though: https://doc.rust-lang.org/1.81.0/src/core/option.rs.html#1135

```rust
pub fn inspect<F: FnOnce(&T)>(self, f: F) -> Self {}
```

not really a big deal, feel free to resolve if intentional

camchenry commented on crates/oxc_linter/src/config/oxlintrc.rs:
```diff
@@ -90,3 +90,49 @@ impl Oxlintrc {
+
+    #[test]
+    fn test_oxlintrc_de_plugins_new() {
+        // let config: Oxlintrc = serde_json::from_value(json!({ "plugins": ["import"] })).unwrap();
```
should we delete this line?

DonIsaac replied:
removed

camchenry commented on crates/oxc_linter/src/options/plugins.rs:
```diff
@@ -151,7 +155,39 @@ impl<S: AsRef<str>> FromIterator<S> for LintPlugins {
+        }
+
+        deserializer.deserialize_any(LintPluginsVisitor)
+        // Vec::<&str>::deserialize_seq(deserializer).map(|vec| vec.into_iter().collect())
```
should we delete this line now?

DonIsaac replied:
removed

Boshen commented on apps/oxlint/fixtures/import/.oxlintrc.json:
```diff
@@ -0,0 +1,7 @@
+{
+    "plugins": ["import"],
```
https://eslint.org/docs/v8.x/use/configure/configuration-files#using-a-configuration-from-a-plugin

The rules for the names are confusing lol:

The eslint-plugin- prefix can be omitted from the plugin name.

DonIsaac replied:
Not only _can_ it be, I've never seen a config file where it wasn't. I don't even think including that prefix is valid. Either way, I didn't add support for it when parsing this field. Should I add it?

Boshen replied:
nah

Boshen reviewed (commented):
I suggest manually test this setup on some of the oxlint-ecosystem-ci projects and see if the rule numbers match.

DonIsaac commented:
We'll need to allow for category-based rule configuration in `oxlintrc.json` files before our next release in order for this PR to be viable. Otherwise there's no way to get that broad-level configuration, followed by rule-by-rule tweaking, that people were getting before using the CLI. Working on a PR now.

DonIsaac commented:
Waiting on #6120 to land before merging this.

DonIsaac commented:
### Merge activity

* **Oct 10, 3:16 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

DonIsaac committed:
feat(linter)!: support plugins in oxlint config files (#6088)

> Closes #5046
This PR migrates the linter crate and oxlint app to use the new `LinterBuilder` API. This PR has the following effects:

1. `plugins` in oxlint config files are now supported
2. irons out weirdness when using CLI args and config files together. CLI args are now always applied on top of config file settings, overriding them.

# Breaking Changes
Before, config files would override rules set in CLI arguments. For example, running this command:
```sh
oxlint -A correctness -c oxlintrc.json
```
With this config file
```jsonc
// oxlintrc.json
{
  "rules": {
    "no-const-assign": "error"
  }
}
```
Would result in a single rule, `no-const-assign` being turned on at an error level with all other rules disabled (i.e. set to "allow").

Now, **CLI arguments will override config files**. That same command with the same config file will result with **all rules being disabled**.

## Details

For a more in-depth explanation, assume we are running the below command using the `oxlintrc.json` file above:
```sh
oxlint -A all -W correctness -A all -W suspicious -c oxlintrc.json
```

### Before
> Note: GitHub doesn't seem to like deeply nested lists. Apologies for the formatting.

Here was the config resolution process _before_ this PR:
<details><summary>Before Steps</summary>

1. Start with a default set of filters (["correctness", "warn"]) if no filters were passed to the CLI. Since some were, the filter list starts out empty.
2. Apply each filter taken from the CLI from left to right. When a filter allows a rule or category, it clears the configured set of rules. So applying those filters looks like this
  a. start with an empty list `[]`
  b. `("all", "allow")` -> `[]`
  c. `("correctness", "warn")` -> `[ <correctness rules> ]`
  d. `("all", "allow")` -> `[]`
  e. `("suspicious", "warn")` -> `[ <suspicious rules> ]`. This is the final rule set for this step
3. Apply overrides from `oxlintrc.json`. This is where things get a little funky, as mentioned in point 2 of what this PR does. At this point, all rules in the rules list are only from the CLI.
  a. If a rule is only set in the CLI and is not present in the config file, there's no effect
  b. If a rule is in the config file but not the CLI, it gets inserted into the list.
  c. If a rule is already in the list and in the config file
    i. If the rule is only present once (e.g. `"no-loss-of-precision": "error"`), unconditionally override whatever was in the CLI with what was set in the config file
    ii. If the rule is present twice (e.g. `"no-loss-of-precision": "off", "@typescript-eslint/no-loss-of-precision": "error"`),
      a. if all rules in the config file are set to `allow`, then turn the rule off
      b. If one of them is `warn` or `deny`, then update the currently-set rule's config object, but _leave its severity alone_.

  So, for our example, the final rule set would be `[<all suspicious rules: "warn">, no-const-assign: "error"]`

</details>

### After
Rule filters were completely re-worked in a previous PR. Now, lint filters aren't kept on hand-only the rule list is.

<details><summary>After Steps</summary>

1. Start with the default rule set, which is all correctness rules for all enabled plugins (`[<all correctness rules: "warn">]`)
2. Apply configuration from `oxlintrc.json` _first_.
  a. If the rule is warn/deny exists in the list already, update its severity and config object. If it's not in the list, add it.
  b. If the rule is set to allow, remove it from the list.

  The list is now `[<all correctness rules except no-const-assign: "warn">, no-const-assign: "error"]`.

3. Apply each filter taken from the CLI from left to right. This works the same way as before. So, after they're applied, the list is now `[<suspicous rules: "warn">]`

</details>
